Fix the promise not fulfilling if the count is unachievable

jstewmon opened the pull request:
Throwing an error when a value is filtered simplifies identifying when
the operation has reached a terminal state and should result in a
rejection.

Another option would be to just remove the `filter` option, since the same thing can be accomplished by the user:
```js
psome([1, 2, 3]
	.map(p => Promise.resolve(p))
	.map(p => p.then(v => {
		if (v < 2) {
			throw new Error('too low');
		}
	})))
```

sindresorhus commented:
> Throwing an error when a value is filtered simplifies identifying when
the operation has reached a terminal state and should result in a
rejection.

Why? It's not clear to me what problem this PR fixes.

> Another option would be to just remove the filter option, since the same thing can be accomplished by the user:

That's not the same thing. The `filter` option is useful as it runs while other promises are running, meaning if the filter returns false, it gives the opportunity to another promise to help fulfill the `count`. So if `count` is 2, and you have 4 promises, it resolves with the two fastest promises, but when filtering afterwards, there's only one left. Had it used the `filter` option, it would have still been 2 promises at the end.

jstewmon commented:
Sorry, I started with the test where the filter function excludes too many values for the promise to fulfill due to a contributing rejection but then I went ahead and worked on a fix. The current implementation will never fulfill if count is unachievable due to a combination of filtered and rejected results.

If you try the new test I added against master, it should fail with a timeout due to being unfulfilled.

jstewmon commented:
My point about the filter option being unnecessary is that a rejection is the same thing as a filter - a rejected result is ignored as long as count resolve. I was trying to show that the input promises (not the p-some result) can be mapped to a continuation that rejects any result that isn’t desired.

sindresorhus commented:
I was going through my PR queue now and realized I never replied here. I'm really sorry about that. If you're still interested in getting this merged, I'm happy to give it the attention it needs.

sindresorhus commented:
I agree with your assessment and the fix here looks good, but I still think the `filter` option is more user-friendly than your `.map` example.

jstewmon commented:
👋 Hi, it's been a while... rebased this on master, so I think it is ready for merging...

This is the test that's failing for node 10:
```js

test('only returns values that passes `filter` option', async t => {
	const f = [
		'foo',
		Promise.resolve(1),
		Promise.resolve('foo'),
		2
	];
	t.deepEqual(await pSome(f, {count: 1, filter: val => typeof val === 'number'}), [1]);
	t.deepEqual(await pSome(f, {count: 2, filter: val => typeof val === 'number'}), [1, 2]);
});
```

I'm not actually sure that first assertion is valid or necessary, since there are two inputs that pass the filter, and which one is returned depends on the resolution order, but I've been testing locally with node 12 and it seems to pass every time.

Edit: figured out the reason it changed is that I removed the unnecessary `await Promise.resolve(element)` wrapping (the wrap was only needed when we did not have `await`, and on Node 10, I think the `await` implementation required an addition promise wrap, so the `Promise.resolve(1)` item takes more cycles to resolve than the literal `2`.

I can make the test pass as-is by adding back the additional `Promise.resolve(element)` wrap, but that seems silly, since it's known to add extra time. I'm going to change the test to use literal `1` and `Promise.resolve(2)`, which I think still demonstrates the same functionality.

sindresorhus commented:
Thank you :)
